File: apps/rust/src/auth/cookies.rs
use axum::http::{header, HeaderMap, HeaderValue};

use super::session::SessionConfig;

pub fn session_cookie_value(config: &SessionConfig, session_id: &str) -> String {
    format!(
        "{}={}; HttpOnly; Path=/; Max-Age={}; SameSite=Lax",
        config.cookie_name, session_id, config.ttl_secs
    )
}

pub fn clear_session_cookie_value(config: &SessionConfig) -> String {
    format!(
        "{}=; HttpOnly; Path=/; Max-Age=0; SameSite=Lax",
        config.cookie_name
    )
}

pub fn append_session_cookie(headers: &mut HeaderMap, config: &SessionConfig, session_id: &str) {
    if let Ok(value) = HeaderValue::from_str(&session_cookie_value(config, session_id)) {
        headers.append(header::SET_COOKIE, value);
    }
}

pub fn append_clear_session_cookie(headers: &mut HeaderMap, config: &SessionConfig) {
    if let Ok(value) = HeaderValue::from_str(&clear_session_cookie_value(config)) {
        headers.append(header::SET_COOKIE, value);
    }
}

pub fn http_access_session_id(headers: &HeaderMap, cookie_name: &str) -> Option<String> {
    session_id_candidates(headers, cookie_name).into_iter().next()
}

pub fn session_id_candidates(headers: &HeaderMap, cookie_name: &str) -> Vec<String> {
    let mut candidates = Vec::new();
    if let Some(value) = headers.get(header::AUTHORIZATION).and_then(|v| v.to_str().ok()) {
        if let Some(token) = parse_bearer(value) {
            candidates.push(token);
        }
    }
    if let Some(value) = headers.get("x-session-id").and_then(|v| v.to_str().ok()) {
        let trimmed = value.trim();
        if !trimmed.is_empty() {
            candidates.push(trimmed.to_string());
        }
    }
    if let Some(cookie) = read_cookie(headers, cookie_name) {
        candidates.push(cookie);
    }
    candidates
}

fn parse_bearer(value: &str) -> Option<String> {
    let trimmed = value.trim();
    let rest = trimmed.strip_prefix("Bearer ")?;
    let token = rest.trim();
    if token.is_empty() {
        None
    } else {
        Some(token.to_string())
    }
}

fn read_cookie(headers: &HeaderMap, name: &str) -> Option<String> {
    let raw = headers.get(header::COOKIE)?.to_str().ok()?;
    for part in raw.split(';') {
        let part = part.trim();
        let Some((key, value)) = part.split_once('=') else {
            continue;
        };
        if key == name {
            let value = value.trim();
            if !value.is_empty() {
                return Some(value.to_string());
            }
        }
    }
    None
}

File: apps/zig/src/auth/cookies.zig
const std = @import("std");
const http = @import("../http_response.zig");
const session_mod = @import("session.zig");

pub fn readSessionCookie(allocator: std.mem.Allocator, request: *std.http.Server.Request, cookie_name: []const u8) ![]const u8 {
    var header_it = request.iterateHeaders();
    while (header_it.next()) |header| {
        if (!std.ascii.eqlIgnoreCase(header.name, "cookie")) continue;
        const raw = std.mem.trim(u8, header.value, " \t\r\n");
        var parts = std.mem.splitScalar(u8, raw, ';');
        while (parts.next()) |part| {
            const trimmed = std.mem.trim(u8, part, " \t");
            if (trimmed.len <= cookie_name.len + 1) continue;
            if (!std.mem.startsWith(u8, trimmed, cookie_name)) continue;
            if (trimmed[cookie_name.len] != '=') continue;
            const value = std.mem.trim(u8, trimmed[cookie_name.len + 1 ..], " \t");
            if (value.len > 0) return try allocator.dupe(u8, value);
        }
    }
    return try allocator.dupe(u8, "");
}

pub fn sessionCookieHeader(allocator: std.mem.Allocator, cookie_name: []const u8, session_id: []const u8, max_age: u32) ![]const u8 {
    return std.fmt.allocPrint(
        allocator,
        "{s}={s}; HttpOnly; Path=/; Max-Age={d}; SameSite=Lax",
        .{ cookie_name, session_id, max_age },
    );
}

pub fn clearSessionCookieHeader(allocator: std.mem.Allocator, cookie_name: []const u8) ![]const u8 {
    return std.fmt.allocPrint(allocator, "{s}=; HttpOnly; Path=/; Max-Age=0; SameSite=Lax", .{cookie_name});
}

pub fn writeSessionJson(
    request: *std.http.Server.Request,
    allocator: std.mem.Allocator,
    config: *const session_mod.SessionConfig,
    session: *const session_mod.SharedSession,
    set_cookie: bool,
) !void {
    const redis_key = try config.redisKey(allocator, session.session_id);
    defer allocator.free(redis_key);
    const body = try session.toResponseJson(allocator, redis_key);
    defer allocator.free(body);
    if (set_cookie) {
        const cookie = try sessionCookieHeader(allocator, config.cookie_name, session.session_id, @intCast(config.ttl_secs));
        defer allocator.free(cookie);
        const extra = [_]std.http.Header{.{ .name = "set-cookie", .value = cookie }};
        return try http.writeJsonResponseExtra(request, .ok, body, &extra);
    }
    return try http.writeJsonResponse(request, .ok, body);
}

pub fn writeSessionCleared(request: *std.http.Server.Request, allocator: std.mem.Allocator, config: *const session_mod.SessionConfig) !void {
    const cookie = try clearSessionCookieHeader(allocator, config.cookie_name);
    defer allocator.free(cookie);
    const extra = [_]std.http.Header{.{ .name = "set-cookie", .value = cookie }};
    return try http.writeNoContentResponse(request, &extra);
}

pub fn extractJsonString(allocator: std.mem.Allocator, json: []const u8, key: []const u8) ?[]const u8 {
    var needle_buf: [64]u8 = undefined;
    const needle = std.fmt.bufPrint(&needle_buf, "\"{s}\"", .{key}) catch return null;
    const key_pos = std.mem.indexOf(u8, json, needle) orelse return null;
    const after_key = json[key_pos + needle.len ..];
    const colon = std.mem.indexOfScalar(u8, after_key, ':') orelse return null;
    var rest = std.mem.trimLeft(u8, after_key[colon + 1 ..], " \t");
    if (rest.len == 0 or rest[0] != '"') return null;
    rest = rest[1..];
    const end = std.mem.indexOfScalar(u8, rest, '"') orelse return null;
    return allocator.dupe(u8, rest[0..end]) catch null;
}
